community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -158,7 +204,18 @@ private W combine( WorkUnit<Material,W> batch )
             {
                 result = result.combine( batch.work );
             }
-            batch = batch.next;
+
+            WorkUnit<Material,W> tmp = batch.next;
+            while ( tmp == null )
+            {
+                // We may see 'null' via race, as work units are put on the
+                // stack before their 'next' pointers are updated. We just spin
+                // until we observe their volatile write to 'next'.
+                // todo Java9: Thread.onSpinWait() ?
+                Thread.yield();
+                tmp = batch.next;
+            }
```

Comment:
Hmm... this is a tricky one. Here is indeed the first place where we are trying to read `WorkUnit.next` now, but it doesn't "feel" like the right place for this spin loop. Accounting for this race would feel more like it "belongs" in the `grabBatch()` method. On the other hand, the longer we can defer the check, the more likely it is we don't have to retry, meaning we should try to defer this look as long as possible - i.e. to here.

Perhaps it would be cleaner in the code to encapsulate this logic in a method, but then it would feel more natural to always use that method for accessing the field, and in `markAsDone()` we don't need to compensate for the race, since `combine()` was invoked before and thus already compensated for it. Although, the overhead of null-checks is really low, so perhaps that's acceptable.